File: C++/String/22_count_occurences.cpp
/*
Idea:
use recursion to traverse through each character of the matrix and taking each character as a start of the string to be found.
if found, repeating for all directions
*/


#include <bits/stdc++.h>
using namespace std;


class Solution {
public:

    int internalSearch(string target, int row, int col, vector<vector<char> >& mat, int row_max, int col_max, int index)
    {
        int found = 0;
        if (row >= 0 && row <= row_max && col >= 0 && col <= col_max && target[index] == mat[row][col])
        {
            // if characters are matched
            char match = target[index];
            index += 1;

            mat[row][col] = 0;

            if (target[index] == 0)
                found = 1; // if string has ended
            else
            {
                // through Backtrack searching in every directions
                found += internalSearch(target, row, col + 1, mat, row_max, col_max, index);
                found += internalSearch(target, row, col - 1, mat, row_max, col_max, index);
                found += internalSearch(target, row + 1, col, mat, row_max, col_max, index);
                found += internalSearch(target, row - 1, col, mat, row_max, col_max, index);
            }
            mat[row][col] = match;
        }
        return found;
    }

    int findOccurrence(vector<vector<char> >& mat, string target)
    {
        int found = 0, row_count = mat.size(), col_count = mat[0].size();

        for (int r = 0; r < row_count; ++r)
            for (int c = 0; c < col_count; ++c)
                found += internalSearch(target, r, c, mat, row_count - 1, col_count - 1, 0);
        return found;
    }
};


int main() {
    int t;
    cin >> t;
    while (t--) {
        int R, C;
        cin >> R >> C;
        vector<vector<char> > mat(R);
        for (int i = 0; i < R; i++) {
            mat[i].resize(C);
            for (int j = 0; j < C; j++) cin >> mat[i][j];
        }
        string target;
        cin >> target;
        Solution obj;
        cout << obj.findOccurrence(mat, target) << endl;
    }
}

File: Python/String/22_count_occurences.py
"""
Idea:
use recursion to traverse through each character of the matrix and taking each character as a start of the string to be found.
if found, repeating for all directions
"""


class Solution:
    def internalSearch(self, target, row, col, mat, row_max, col_max, index):
        found = 0
        if 0 <= row <= row_max and 0 <= col <= col_max and target[index] == mat[row][col]:
            # If characters are matched
            match = target[index]
            index += 1

            mat[row][col] = 0

            if target[index] == 0:
                found = 1  # If the string has ended
            else:
                # Through backtracking, search in every direction
                found += self.internalSearch(target, row, col + 1, mat, row_max, col_max, index)
                found += self.internalSearch(target, row, col - 1, mat, row_max, col_max, index)
                found += self.internalSearch(target, row + 1, col, mat, row_max, col_max, index)
                found += self.internalSearch(target, row - 1, col, mat, row_max, col_max, index)
            mat[row][col] = match
        return found

    def findOccurrence(self, mat, target):
        found = 0
        row_count = len(mat)
        col_count = len(mat[0])

        for r in range(row_count):
            for c in range(col_count):
                found += self.internalSearch(target, r, c, mat, row_count - 1, col_count - 1, 0)
        return found


# Initial Template for Python 3
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        line = input().strip().split()
        R = int(line[0])
        C = int(line[1])
        mat = []
        for _ in range(R):
            mat.append([x for x in input().strip().split()])
        target = input()
        obj = Solution()
        print(obj.findOccurrence(mat, target))
# } Driver Code Ends
